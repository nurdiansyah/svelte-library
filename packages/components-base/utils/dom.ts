export const ownerDocument = (node: Node) => (node && node.ownerDocument) || document;
export const ownerWindow = (node: Node) => {
  const doc = ownerDocument(node);
  return doc.defaultView || window;
};

// A change of the browser zoom change the scrollbar size.
// Credit https://github.com/twbs/bootstrap/blob/3ffe3a5d82f6f561b82ff78d82b32a7d14aed558/js/src/modal.js#L512-L519
export const getScrollbarSize = () => {
  const scrollDiv = document.createElement("div");
  scrollDiv.style.width = "99px";
  scrollDiv.style.height = "99px";
  scrollDiv.style.position = "absolute";
  scrollDiv.style.top = "-9999px";
  scrollDiv.style.overflow = "scroll";

  document.body.appendChild(scrollDiv);
  const scrollbarSize = scrollDiv.offsetWidth - scrollDiv.clientWidth;
  document.body.removeChild(scrollDiv);

  return scrollbarSize;
};

export const getOriginalBodyPadding = () => {
  const style = window ? window.getComputedStyle(document.body, null) : { getPropertyValue: (e?: any) => undefined };

  return parseInt((style && style.getPropertyValue("padding-right")) || "0", 10);
};

export const setScrollbarWidth = (padding) => {
  if (padding > 0) {
    document.body.style.paddingRight = `${padding}px`;
  }
};

export const isBodyOverflowing = () => (window ? document.body.clientWidth < window.innerWidth : false);
