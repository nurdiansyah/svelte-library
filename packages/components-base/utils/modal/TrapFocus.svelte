<script context="module" lang="ts">
  export interface Props {
    disableAutoFocus: boolean;
    disableEnforceFocus: boolean;
    disableRestoreFocus: boolean;
    node: HTMLElement | undefined;
    enable: boolean;
    getDoc: () => Document;
    open: boolean;
    onFocus?: (e?: Event) => void;
    onTabChange?: (e?: KeyboardEvent) => void;
  }

  interface Options extends Omit<Props, "node"> {
    sentinelStart: HTMLElement;
    sentinelEnd: HTMLElement;
  }

  function trapFocus(node: HTMLElement, options: Options) {
    let ignoreNextEnforceFocus: boolean;
    let restoreNode: HTMLElement | null;
    let doc: Document;

    function saveToRestore() {
      restoreNode = !open || typeof window == "undefined" ? null : (doc.activeElement as HTMLElement);
    }
    // With Edge, Safari and Firefox, no focus related events are fired when the focused area stops being a focused area
    // e.g. https://bugzilla.mozilla.org/show_bug.cgi?id=559561.
    //
    // The whatwg spec defines how the browser should behave but does not explicitly mention any events:
    // https://html.spec.whatwg.org/multipage/interaction.html#focus-fixup-rule.
    const interval = setInterval(() => {
      contain();
    }, 50);

    function contain(e?: Event) {
      if (options.disableEnforceFocus || !options.enable || ignoreNextEnforceFocus) {
        ignoreNextEnforceFocus = false;
        return;
      }

      if (!node.contains(doc.activeElement)) {
        node.focus();
      }

      options.onFocus && options.onFocus(e);
    }

    function loopFocus(e: KeyboardEvent) {
      // 9 = tab
      if (options.disableEnforceFocus || !options.enable || e.code !== "9") {
        return;
      }

      // Make sure the next tab starts from the right place.
      if (doc.activeElement === node) {
        // We need to ignore the next contain as
        // it will try to move the focus back to the rootRef element.
        ignoreNextEnforceFocus = true;
        if (e.shiftKey) {
          options.sentinelEnd.focus();
        } else {
          options.sentinelStart.focus();
        }
      }
      options.onTabChange && options.onTabChange(e);
    }

    function effect() {
      if (!options.open) {
        return;
      }

      if (!options.disableRestoreFocus) {
        saveToRestore();
      }

      if (!options.disableAutoFocus && node.contains(doc.activeElement)) {
        if (!node.hasAttribute("tabIndex")) {
          if (process.env.NODE_ENV !== "production") {
            console.error(
              [
                "The modal content node does not accept focus.",
                'For the benefit of assistive technologies, the tabIndex of the node is being set to "-1".'
              ].join("\n")
            );
          }
          node.setAttribute("tabIndex", "-1");
        }
        node.focus();
      }

      doc.addEventListener("focus", contain, true);
      doc.addEventListener("keydown", loopFocus, true);
    }

    return {
      update(_options: Options) {
        options = _options;
        doc = options.getDoc();
        effect();
      },
      destroy() {
        clearInterval(interval);
        doc.removeEventListener("focus", contain);
        doc.removeEventListener("keydown", loopFocus);

        // restoreLastFocus()
        if (!options.disableRestoreFocus) {
          // In IE 11 it is possible for document.activeElement to be null resulting
          // in nodeToRestore.current being null.
          // Not all elements in IE 11 have a focus method.
          // Once IE 11 support is dropped the focus() call can be unconditional.
          if (restoreNode && restoreNode.focus) {
            restoreNode.focus();
          }

          restoreNode = null;
        }
      }
    };
  }
</script>

<script lang="ts">
  import { ownerDocument } from "../dom";
  import { onMount, onDestroy } from "svelte";

  // state
  let sentinelStart: HTMLElement;
  let sentinelEnd: HTMLElement;

  // props
  export let disableAutoFocus: boolean = false;
  export let disableEnforceFocus: boolean = false;
  export let disableRestoreFocus: boolean = false;
  export let onFocus: ((e?: Event) => void) | undefined = undefined;
  export let onTabChange: ((e?: KeyboardEvent) => void) | undefined = undefined;
  export let node: HTMLElement | undefined = undefined;
  export let enable: boolean = true;
  export let getDoc: () => Document = () => ownerDocument(node);
  export let open: boolean;
</script>

<div
  {...$$restProps}
  bind:this={node}
  use:trapFocus={{ disableAutoFocus, getDoc, open, disableEnforceFocus, disableRestoreFocus, enable, sentinelEnd, sentinelStart, onFocus, onTabChange }}>
  <div bind:this={sentinelStart} tabindex={0} data-testid="sentinel-start" />
  <slot />
  <div bind:this={sentinelEnd} tabindex={0} data-testid="sentinel-end" />
</div>
