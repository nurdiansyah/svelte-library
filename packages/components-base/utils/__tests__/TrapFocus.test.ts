import { render as renderSvelte, act, fireEvent, cleanup } from "@testing-library/svelte";
import userEvent from "@testing-library/user-event";
import { ownerDocument } from "../dom";
import TrapFocus from "../modal/TrapFocus.svelte";

describe("<TrapFocus />", () => {
  const testId = "trap-focus";
  let props = {
    "data-testid": testId
  };
  const render = (_props = {}) => {
    _props = Object.assign(props, _props);
    return renderSvelte(TrapFocus, props, { container: document.body });
  };

  describe("props", () => {
    afterEach(() => {
      cleanup();
    });
    it("check: open", () => {
      const { getByTestId, container } = render({ open: true });
      let el: undefined | HTMLElement;
      expect(() => (el = getByTestId(testId))).not.toThrow();
      expect(el).toBeInTheDocument();
      expect(container).toMatchSnapshot();
    });
    it("check: enable", async () => {
      const { getByTestId, component } = render({ open: true });
      let el = getByTestId(testId);
      expect(el).toBeInTheDocument();
      await act(() => {
        component.$set({ enable: false });
      });
      cleanup();
    });

    describe("event", () => {
      const onTabChange = (e: KeyboardEvent) => {
        if (e.code === "9") {
        }
      };
      it("check: keydown", async () => {
        const { getByTestId } = render({ open: true, onTabChange });
        let el = getByTestId(testId);
        let sentinelStart = getByTestId("sentinel-start");
        let sentinelEnd = getByTestId("sentinel-end");
        const doc = ownerDocument(el);
        expect(doc.activeElement).not.toEqual(el);
        userEvent.tab();
        await fireEvent.keyDown(el, { code: 9 });
        expect(doc.activeElement).toEqual(sentinelStart);
        userEvent.tab();
        await fireEvent.keyDown(el, { code: 9 });
        expect(doc.activeElement).toEqual(sentinelEnd);
        // let isFocus = false;
        // el.addEventListener("focusin", () => {
        //   isFocus = true;
        // });
        // await fireEvent.focusOut(el);
      });
      it("check: focus child active when autofocus", async () => {
        const { getByTestId, component } = renderSvelte(TrapFocus, {
          props: {
            ...props,
            open: true,
            onTabChange,
            disableAutoFocus: true
          }
        });
        let el = getByTestId(testId);
        let sentinelStart = getByTestId("sentinel-start");
        expect(el.hasAttribute("tabIndex")).toBeFalsy();
        sentinelStart.focus();
        expect(el.tabIndex).toBe(-1);
        await act(() => {
          component.$set({ disableAutoFocus: false });
        });
        expect(el.hasAttribute("tabIndex")).toBeTruthy();
        expect(el.tabIndex).toBe(-1);

        el = getByTestId(testId);
        expect(el.tabIndex).toBe(-1);
      });
    });
  });
});
