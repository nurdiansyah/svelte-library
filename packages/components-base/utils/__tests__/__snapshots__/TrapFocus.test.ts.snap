// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TrapFocus /> props check: open 1`] = `
<body>
  <div>
    <div
      data-testid="trap-focus"
    >
      <div
        data-testid="sentinel-start"
        tabindex="0"
      />


      <div
        data-testid="sentinel-end"
        tabindex="0"
      />
    </div>
  </div>
</body>
`;
